server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerRecipientRateLimitMailet.scala

```diff
@@ -79,20 +80,53 @@ class PerRecipientRateLimitMailet @Inject()(rateLimiterFactoryProvider: RateLimi
 
   override def service(mail: Mail): Unit = {
     if (!mail.getRecipients.isEmpty) {
-      val mailAddressAndRateLimitResults: Seq[(MailAddress, RateLimitingResult)] =
-        mail.getRecipients.asScala.map(recipient => (recipient, applyRateLimiter(mail, recipient))).toSeq
-
-      if (allResultsAreExceeded(mailAddressAndRateLimitResults.map(_._2))) {
-        mail.setState(exceededProcessor)
-      } else {
-        mailAddressAndRateLimitResults.foreach {
-          case (mailAddress: MailAddress, rateLimitingResult: RateLimitingResult) =>
-            val newMail: MailImpl = MailImpl.duplicate(mail)
-            newMail.setRecipients(ImmutableList.of(mailAddress))
-            if (rateLimitingResult.equals(RateExceeded)) {
-              newMail.setState(exceededProcessor)
-            }
-            getMailetContext.sendMail(newMail)
+      val rateLimitResults: Seq[(MailAddress, RateLimitingResult)] = mail.getRecipients.asScala.toSeq
+        .map(recipient => (recipient, applyRateLimiter(mail, recipient)))
+
+      exceededProcess(mail, rateLimitResults)
+      acceptableProcess(mail, rateLimitResults)
+    }
+  }
+
+  private def exceededProcess(mail: Mail, rateLimitResults: Seq[(MailAddress, RateLimitingResult)]): Unit = {
+    // if all recipients exceed the rate limit -> overwrite the state mail property
+    // else cloning the mail (with exceeded recipients) and send the clone to the exceededProcessor
+    if (!rateLimitResults.exists(_._2.equals(AcceptableRate))) {
+      mail.setState(exceededProcessor)
+    } else {
+      val exceededRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(RateExceeded)).map(_._1)
+
+      if (exceededRecipients.nonEmpty) {
+        mail.setState(Mail.GHOST)
+        val newMail: MailImpl = MailImpl.duplicate(mail)
+        try {
+          newMail.setRecipients(ImmutableList.copyOf(exceededRecipients.asJava))
+          getMailetContext.sendMail(newMail, exceededProcessor)
+        } finally {
+          LifecycleUtil.dispose(newMail)
+        }
+      }
+    }
+  }
+
+  private def acceptableProcess(mail: Mail, rateLimitResults: Seq[(MailAddress, RateLimitingResult)]): Unit = {
+    // if all recipients acceptable the rate limit -> do nothing
+    // else cloning the mail (with acceptable recipients) and send the clone to the next processor
+    if (!rateLimitResults.exists(_._2.equals(RateExceeded))) {
+      // do nothing
+    }
+    else {
+      val acceptableRecipients: Seq[MailAddress] = rateLimitResults.filter(_._2.equals(AcceptableRate)).map(_._1)
+
+      if (acceptableRecipients.nonEmpty) {
+        mail.setState(Mail.GHOST)
+        val newMail: MailImpl = MailImpl.duplicate(mail)
+        try {
+          newMail.setRecipients(ImmutableList.copyOf(acceptableRecipients.asJava))
+          // state = "transport" to avoid loop
+          getMailetContext.sendMail(newMail, Mail.TRANSPORT)
```

Comment:
after checking rate_limiter, it needs to be crossed to the next step. (avoid the loop of rate-limiting checking).
This means it needs to be placed `NOT` in "transport" state, and before "transport".